### DISCUSSION TITLE
Shared predictors between series

### DISCUSSION THREAD

**Discussion by AsierRGdD** [Nixtla/mlforecast#83] [Q&A]
Hi,

I'm such trouble with one problem and I don't know if mlforecast can deal with it.

As in the readme appears, there are some static features that can be given to the model in order to have more information i.e **static_0**:

ds | y | static_0
-- | -- | --
2000-01-01 | 1.751917 | 72
2000-01-02 | 9.196715 | 72
2000-01-03 | 18.577788 | 72
2000-01-04 | 24.520646 | 72
2000-01-05 | 33.418028 | 72

My question is divided in two parts:
a) Can the static value be categorical?
b) Can different series 'share' this values as extra predictors in order to learn patterns from them? For example:

id | ds | y | pred_0 | pred_1 | pred_2 
-- | -- | -- | -- | -- | -- 
1 |2000-01-01 | 1.751917 | 'cat0_A' | 'cat1_B' | 'cat2_A'
1 |2000-01-02 | 9.196715 | 'cat0_A' | 'cat1_B' | 'cat2_A'
1 |2000-01-03 | 18.577788  | 'cat0_A' | 'cat1_B' | 'cat2_A'
2 |2000-01-02 | 34.520646  | 'cat0_A' | 'cat1_A' | 'cat2_A'
2 |2000-01-03 | 43.418028  | 'cat0_A' | 'cat1_A' | 'cat2_A'
2 |2000-01-04 | 24.520646  | 'cat0_A' | 'cat1_A' | 'cat2_A'
2 |2000-01-05 | 33.418028  | 'cat0_A' | 'cat1_A' | 'cat2_A'

As you can see in these dataset example, id series 1 and 2 share pred_0 and pred_2 within two days (02 and 03) so there might be a pattern under these predictors which affects for both series. Does mlforecast handle this kind of problem?  If not, any hint or advice for solving it? 

Thanks a lot

**Comment by jmoralez** [comment]
Hi @AsierRGdD, the answer to both questions is yes, you just need to pass them in your dataframe to the fit methods and use a model that can use them (e.g. LightGBM). If you don't specify a value for `static_features` they will all be interpreted as static. Please let us know if you have further doubts, also you might find the [docs](https://nixtla.github.io/mlforecast/) useful if you want to dig deeper.

**Comment by AsierRGdD** [comment]
Hi @jmoralez ! Thanks for the quick reply, it has been really valuable for me. I still don't understand how to forecast with a certain combination of these shared predictors. Do I have to include [pred_0 | pred_1 | pred_2] as _data_  parameter in fit whereas in forecasting it would be passed to _new_data_? 

I only have another little question around the data and the possible regressor (i.e LightGBM):

How is going to work the model with low data (more or less 20 days for each id with shared predictors). I have been thinking also in any kind of bayesian method.

Thanks again

**Reply by jmoralez** [reply]
> Do I have to include [pred_0 | pred_1 | pred_2] as data parameter in fit whereas in forecasting it would be passed to new_data?

If the features don't change over time you just need to make sure they're included in your data argument in fit and when calling predict they will be repeated automatically.

> How is going to work the model with low data

It depends on the features you use and how many ids you have, but probably it won't be that good.

**Comment by davidrpugh** [comment]
I also am a bit confused by how `mlforecast` uses `static_features`. Consider the [Store Sales Time Series Forecasting](https://www.kaggle.com/competitions/store-sales-time-series-forecasting) competition on Kaggle. There are a large number of static features for this problem: `store_id` and `family` (which together make up the `unique_id` for a series) but also other variables like `city`, `state`, `store_type`, and `store_cluster` are all variables that are fixed over time for an individual series but vary across series.

My understanding is the `mlforecast` fits models to individual series based on `unique_id`. Since `static_features` don't vary within a series, how can they be useful in forecasting `y` for a series? Does mlforecast also fit models globally using the whole dataset and then use the global model with the individual models when forecasting values for individual series?

Perhaps I have it backwards and `mlforecast` fits a global model using the whole dataset comprised of all `unique_id` and then uses that model to produce forecasts of `y` for each individual  `unique_id` series.

**Reply by jmoralez** [reply]
Hey @davidrpugh

> mlforecast fits a global model using the whole dataset comprised of all unique_id and then uses that model to produce forecasts of y for each individual unique_id series.

that's exactly what it does. So having those static features allows the models to learn common patterns by city for example across different ids.
